module/trie.c

```diff
@@ -0,0 +1,286 @@
+    }
+    size_t dummy;
+    char *temp = RedisModule_StringPtrLen(argv[2], &dummy);
+    char *empty = "";
+    if (temp == NULL || strcmp(temp, empty) == 0) {
+    	return RedisModule_ReplyWithError(ctx,"ERR invalid value: must be a string");
+    } 
+
+    struct trie *t;
+    /* Create an empty value object if the key is currently empty. */
+    if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    	t = new_trie('\0');
+    	RedisModule_ModuleTypeSetValue(key,trie,t);
+    } else {
+        t = RedisModule_ModuleTypeGetValue(key);
+    }
+
+    /* Insert the new string. */
+    insert_string(temp, t);
+    
+	RedisModule_ReplyWithSimpleString(ctx, "Success");    
+	RedisModule_ReplicateVerbatim(ctx);
+    return REDISMODULE_OK;
+}
+
+/* TRIE.CONTAINS key value */
+int TrieContains_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx); /* Use automatic memory management. */
+
+    if (argc != 3) return RedisModule_WrongArity(ctx);
+
+    RedisModuleKey *key = RedisModule_OpenKey(ctx,argv[1],
+        REDISMODULE_READ|REDISMODULE_WRITE);
+    int type = RedisModule_KeyType(key);
+    if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    	return RedisModule_ReplyWithError(ctx,"ERR invalid key: not an existing trie");
+    }
+    else if (RedisModule_ModuleTypeGetType(key) != trie)
+    {
+        return RedisModule_ReplyWithError(ctx,REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
+    size_t dummy;
+    char *temp = RedisModule_StringPtrLen(argv[2], &dummy);
+
+    struct trie *t;
+    t = RedisModule_ModuleTypeGetValue(key);
+
+    /* Check for the string. */
+    int c = trie_search(temp, t);
+    
+    if (c == 1) {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie contains the word.");
+    } else if (c == 0) {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie does not contain the word.");
+    } else {
+        RedisModule_ReplyWithSimpleString(ctx, "The trie contains it as a prefix but not as a word.");  
+    }       
+    RedisModule_ReplicateVerbatim(ctx);
+    return REDISMODULE_OK;
+}
+
+/* ========================== "trie" type methods (Redis data saving and entry functions) ======================= */
+
+/* This function must be present on each Redis module. It is used in order to
+ * register the commands into the Redis server. */
+
+int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    REDISMODULE_NOT_USED(argv);
+    REDISMODULE_NOT_USED(argc);
+
+    if (RedisModule_Init(ctx,"trie123az",1,REDISMODULE_APIVER_1)
+        == REDISMODULE_ERR) return REDISMODULE_ERR;
+
+    RedisModuleTypeMethods tm = {
+        .version = REDISMODULE_TYPE_METHOD_VERSION,
+        .rdb_load = NULL,
+        .rdb_save = NULL,
+        .aof_rewrite = NULL,
+        .mem_usage = NULL,
+        .free = NULL,
+        .digest = NULL
+    };
+
+    trie = RedisModule_CreateDataType(ctx,"trie123az",0,&tm);
```

Comment:
I think this is the origin of the `struct trie` confusion above. There needs to be a comment here explaining exactly what this is doing. Where was `trie` declared? Is this referencing `struct trie`? This is not at all clear.

Reply:
You're definitely right that this is the source of all the confusion above. So in this file, there are two different tries. You have the struct trie, which is the low level C trie. And then you have the Redis trie data structure which is just called trie. Look at line 16 "static RedisModuleType *trie". This declares a global variable that stores the reference to the Redis trie data type. The function, RedisModule_CreateDataType, registers a new data type that's exported by the module (and that reference is stored in the variable created in line 16). Think of this as sort of a wrapper to the low level struct trie data type. You need this wrapper to directly interact with the Module commands which is why the transition is made from using struct trie (in the first part of the module that I labeled as Internal data structure) to trie (in the second part of the module that I labeled as trie type commands). Think of this second part of the module as the outward facing commands that actually interact with the user and the Redis storage of data structures. So naturally, the actual data structure you're interacting with on the Redis level isn't struct trie, but just "trie" that you registered with the call to RedisModule_CreateDataType.

To address you and Valerie's specific concerns, in line 193 (" RedisModule_ModuleTypeGetType(key) != trie) RedisModule_Module_TypeGetType actually returns the Redis data structure and not the low level one, so the comparison to trie is valid. In line 208 "RedisModule_ModuleTypeSetValue(key,trie,t)" RedisModule_ModuleTypeSetValue actually sets the value of the key (which previously has no value associated with it) to the Redis data type of the second argument and stores the third argument in that redis value (expects a void*). Line 228 can be explained in the same way as line 193.  @valeriehan6  as well to see this explanation

Reply:
Awesome, this is a great explanation, really thorough. Thanks :+1: